File: examples/node/batch-send.js
'use strict';

// Send thousands of SMS by chunking the list and calling sendBulk() per chunk.
//
// Run:  PUSHFY_API_TOKEN=... node batch-send.js
//
// Strategy:
//   - Split the audience into fixed-size chunks (e.g. 500) — one API call each.
//   - Give every message a stable, unique extId (here derived from the row id)
//     so the whole batch is safe to re-run after a partial failure.
//   - Add a small pause between chunks to stay under rate limits.

const { Pushfy, RateLimitError } = require('@pushfy/pushfy');

const pushfy = new Pushfy({ apiToken: process.env.PUSHFY_API_TOKEN });

const CHUNK_SIZE = 500;
const PAUSE_MS = 250;
const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

function* chunk(arr, size) {
  for (let i = 0; i < arr.length; i += size) yield arr.slice(i, i + size);
}

// Pretend this comes from your database. Each recipient has a stable id.
function buildAudience(n) {
  const rows = [];
  for (let i = 1; i <= n; i++) {
    rows.push({
      id: i,
      to: '5511999999999',
      text: `Reminder #${i}: your appointment is tomorrow`,
      extId: `reminder-2026-07-13-${i}`, // stable → idempotent re-run
    });
  }
  return rows;
}

async function main() {
  const audience = buildAudience(2500);
  let sent = 0;

  for (const batch of chunk(audience, CHUNK_SIZE)) {
    const payload = batch.map((r) => ({ to: r.to, text: r.text, extId: r.extId }));

    // Retry THIS chunk on rate limits — same extIds make the resend idempotent.
    for (;;) {
      try {
        const res = await pushfy.sms.sendBulk(payload);
        sent += res.length;
        console.log(`Chunk accepted: ${res.length} (total ${sent}/${audience.length})`);
        break;
      } catch (err) {
        if (err instanceof RateLimitError) {
          const wait = (err.retryAfter || 2) * 1000;
          console.warn(`Rate limited; waiting ${wait}ms then retrying chunk`);
          await sleep(wait);
          continue;
        }
        throw err; // 4xx/auth: fix the payload — do not loop forever
      }
    }
    await sleep(PAUSE_MS);
  }

  console.log(`Done. ${sent} message(s) accepted.`);
}

main().catch((err) => {
  console.error('Batch failed:', err.status, err.code, err.message);
  process.exitCode = 1;
});

File: examples/python/batch_send.py
"""Send a large recipient list by splitting it into chunks.

Run:  PUSHFY_API_TOKEN=... python3 batch_send.py

A single request accepts up to 100,000 messages, but smaller chunks keep each
request well under the payload limit, give steadier throughput, and make retries
cheaper. Every message keeps its own unique ext_id so failures can be retried
per-chunk without duplicating the successful ones.
"""

import os

from pushfy import Pushfy, ApiError, RateLimitError

CHUNK_SIZE = 1000


def chunked(items, size):
    for i in range(0, len(items), size):
        yield items[i:i + size]


def build_messages():
    # Stand-in recipient list. Replace with your real audience.
    base = 5511999990000
    return [
        {
            "to": str(base + i),
            "text": "Hello from Pushfy",
            "ext_id": "batch-%06d" % i,
        }
        for i in range(2500)
    ]


def main():
    pushfy = Pushfy(api_token=os.environ["PUSHFY_API_TOKEN"])
    messages = build_messages()

    accepted = 0
    failed_chunks = []

    for index, chunk in enumerate(chunked(messages, CHUNK_SIZE)):
        try:
            pushfy.sms.send_bulk(chunk)
            accepted += len(chunk)
            print("Chunk %d ok (%d messages)" % (index, len(chunk)))
        except (RateLimitError, ApiError) as err:
            # Record the chunk; retry it later (see retry.py) — the stable
            # ext_ids make a re-send safe.
            print("Chunk %d failed: %s" % (index, type(err).__name__))
            failed_chunks.append(index)

    print("Done. Accepted %d/%d messages, %d chunk(s) to retry."
          % (accepted, len(messages), len(failed_chunks)))


if __name__ == "__main__":
    main()
